merkle-tree/hasher/src/bytes.rs

```diff
@@ -0,0 +1,217 @@
+        match self {
+            Some(hashable) => {
+                let mut bytes = hashable.as_byte_vec();
+                bytes.reserve(1);
+                bytes.insert(0, vec![1]);
+                bytes
+            }
+            None => vec![vec![0]],
+        }
+    }
+}
+
+impl<const N: usize> AsByteVec for [u8; N] {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.to_vec()]
+    }
+}
+
+impl AsByteVec for String {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.as_bytes().to_vec()]
+    }
+}
+
+#[cfg(feature = "solana")]
+impl AsByteVec for solana_program::pubkey::Pubkey {
+    fn as_byte_vec(&self) -> Vec<Vec<u8>> {
+        vec![self.to_bytes().to_vec()]
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_as_byte_vec_integers() {
+        let i8_min: &dyn AsByteVec = &i8::MIN;
+        let i8_min_bytes = i8_min.as_byte_vec();
+        assert_eq!(i8_min_bytes, &[&[128]]);
+        assert_eq!(i8_min_bytes, &[i8::MIN.to_ne_bytes()]);
+        let i8_max: &dyn AsByteVec = &i8::MAX;
+        let i8_max_bytes = i8_max.as_byte_vec();
+        assert_eq!(i8_max_bytes, &[&[127]]);
+        assert_eq!(i8_max_bytes, &[i8::MAX.to_ne_bytes()]);
+
+        let u8_min: &dyn AsByteVec = &u8::MIN;
+        let u8_min_bytes = u8_min.as_byte_vec();
+        assert_eq!(u8_min_bytes, &[&[0]]);
+        assert_eq!(u8_min_bytes, &[u8::MIN.to_ne_bytes()]);
+        let u8_max: &dyn AsByteVec = &u8::MAX;
+        let u8_max_bytes = u8_max.as_byte_vec();
+        assert_eq!(u8_max_bytes, &[&[255]]);
+        assert_eq!(u8_max_bytes, &[u8::MAX.to_ne_bytes()]);
+
+        let i16_min: &dyn AsByteVec = &i16::MIN;
+        let i16_min_bytes = i16_min.as_byte_vec();
+        assert_eq!(i16_min_bytes, &[&[0, 128]]);
+        assert_eq!(i16_min_bytes, &[&i16::MIN.to_ne_bytes()]);
+        let i16_max: &dyn AsByteVec = &i16::MAX;
+        let i16_max_bytes = i16_max.as_byte_vec();
+        assert_eq!(i16_max_bytes, &[&[255, 127]]);
+        assert_eq!(i16_max_bytes, &[i16::MAX.to_ne_bytes()]);
+
+        let u16_min: &dyn AsByteVec = &u16::MIN;
+        let u16_min_bytes = u16_min.as_byte_vec();
+        assert_eq!(u16_min_bytes, &[&[0, 0]]);
+        assert_eq!(u16_min_bytes, &[u16::MIN.to_ne_bytes()]);
+        let u16_max: &dyn AsByteVec = &u16::MAX;
+        let u16_max_bytes = u16_max.as_byte_vec();
+        assert_eq!(u16_max_bytes, &[&[255, 255]]);
+        assert_eq!(u16_max_bytes, &[u16::MAX.to_ne_bytes()]);
+
+        let i32_min: &dyn AsByteVec = &i32::MIN;
+        let i32_min_bytes = i32_min.as_byte_vec();
+        assert_eq!(i32_min_bytes, &[&[0, 0, 0, 128]]);
+        assert_eq!(i32_min_bytes, &[i32::MIN.to_ne_bytes()]);
+        let i32_max: &dyn AsByteVec = &i32::MAX;
+        let i32_max_bytes = i32_max.as_byte_vec();
+        assert_eq!(i32_max_bytes, &[&[255, 255, 255, 127]]);
+        assert_eq!(i32_max_bytes, &[i32::MAX.to_ne_bytes()]);
```

Comment:
1) aren't these all redundant assertions? 
2) provided that `as_byte_vec` returns a vec, shouldn't the assertion be against `&[vec![255, 255, 255, 127]]` ?
@vadorovsky

Reply:
1. Well, in a way they are, but I just wanted to assure myself that `as_byte_vec` on primitives returns the same values as the std functions, but also explicitly check the exact values.
2. In Rust, you can compare vectors with slices, so these assertions work. :)